### PR TITLE
Think masking

### PR DESCRIPTION
**Pull request by divya-kumari32**
This PR adds support for optionally appending `\n<think>\n` to the default `asst_tag` during span-based label masking for SFT — but only for samples that contain `<think>` content and when the user explicitly specifies it in the `mask_think_tag` flag. This feature is designed to additionally support models that are  **_thinking_** models, but do not specify *thinking mode* by default, and also in case the `think` and `no-think` samples are mixed in the dataset during training. 

Some assistant responses in a sample might contain a `<think>` block that represents internal monologue/reasoning by the model. During SFT, the user might not want the model to learn to *generate* the `<think>` tag itself. Instead, we optionally allow `<think>` to be treated as part of the assistant tag span boundary, so masking cleanly includes or excludes it.

The behavior is controlled by two conditions:

1. `mask_think_tag` (passed by the user)
   - `False`: masking ignores `<think>` completely.  
     - This may be because the user doesn’t want `<think>` included,  
       the model is not a “thinking” model, or for other design/training reasons.  
   - `True`: masking *may* append `<think>` to the assistant tag, depending on the passed sample's content.

2. `has_thinking_content(messages)` (sample check) 
   - Returns `True` if the assistant message:  
     - Has an explicit `thought` field, **or**  
     - Contains `"<think>"` inside its content string. 

The `<think>` suffix is appended to the assistant tag **only when both are true**:

```python
if mask_think_tag and has_thinking_content(messages):
   asst_tag += think_tag
```

**Additional notes**: 
Other minor changes:
Although it is common for `add_special_tokens` to be set to False during tokenizer encoding, we set `add_special_tokens=False` here to disable the injection of tokenizer-specific BOS/EOS or SEP tokens that would otherwise be automatically added to the encoded sequence. This is essential for correct behavior in tasks like:
- Span matching or pattern matching on tokenized control tags (e.g., <|start_of_role|>assistant<|end_of_role|>)
- Label masking for SFT or instruction tuning pipelines, where precise position tracking is required
Why is this necessary:
- By default, tokenizers for models like LLaMA, Mistral, and Phi-2 have `add_bos_token=True`, which means they automatically prepend a BOS token. This would shift all token positions by 1 and break any span matching logic.
- Other models like Qwen, etc., have `add_bos_token=False` by default, so they behave differently — leading to inconsistent behavior across model families if not explicitly handled.

### CODE REVIEW

File: open_instruct/dataset_transformation.py [hunk 1]
```diff
@@ -1178,26 +1178,45 @@ def sft_span_seach_mask_out(
     row: Dict[str, Any],
     tokenizer: PreTrainedTokenizer,
     max_seq_length: int,
-    asst_tag: str = "<|start_of_role|>assistant<|end_of_role|>",
```
Comment (fabianlim) [inline open_instruct/dataset_transformation.py:1181]: I dont understand why this is removed.. we should not destroy backward compatibility if we are reusing this `sft_span_seach_mask_out` function. we can always create other function if the functionality changed too much.

File: open_instruct/dataset_transformation.py [hunk 2]
```diff
@@ -1178,26 +1178,45 @@ def sft_span_seach_mask_out(
...
+        else "<|start_of_role|>assistant<|end_of_role|>"
+    )
+
     def masking_strategy_span_search(input_ids: torch.tensor, tokenizer):
         # some prep
         match = lambda x, y: torch.all(x == y)
-        _asst_tag = tokenizer.encode(asst_tag)
-        _end_tag = tokenizer.encode(end_tag)
+        # `asst_tag` is now captured from the outer scope's dynamic variable
+        _asst_tag = tokenizer.encode(asst_tag, add_special_tokens=False)
```
Comment (fabianlim) [inline open_instruct/dataset_transformation.py:1218]: there is code duplication here

File: open_instruct/dataset_transformation.py (hunk 1 above)
Comment (divya-kumari32) [reply open_instruct/dataset_transformation.py:1181]: Got it, thanks missed that

File: open_instruct/dataset_transformation.py [hunk 3]
```diff
@@ -1178,26 +1178,45 @@ def sft_span_seach_mask_out(
...
+    It dynamically determines the assistant tag based on the presence of a
+    <think> block in the assistant's response.
+    """
+    
+    messages = row["messages"]
+    if len(messages) == 0:
+        raise ValueError("messages field is empty.")
+
+    # Dynamically determine the assistant tag based on the conversation content.
+    is_think_sample = False
```
[the remark this answers is not in the document]
Comment (fabianlim) [reply open_instruct/dataset_transformation.py:1198]: sorry that was not what I meant. Just something like this

```
def is_think(messages):
  for message in messages:
          if message.get("role") == "assistant":
              # Check for an explicit 'thought' field or a '<think>' tag in the content.
              if message.get("thought") or (
                  isinstance(message.get("content"), str) and "<think>" in message["content"]
              ):
                  return True # A single 'think' block defines the sample type.
```

File: open_instruct/dataset_transformation.py (hunk 2 above)
Comment (fabianlim) [reply open_instruct/dataset_transformation.py:1219]: ok sorry i misread. But im not sure why we need `add_special_tokens`.

File: open_instruct/dataset_transformation.py [hunk 4]
```diff
@@ -1233,7 +1243,11 @@ def masking_strategy_span_search(input_ids: torch.tensor, tokenizer):
     additional_inputs = {}
     for k in ["tools", "documents"]:
         if k in row:
-            additional_inputs[k] = row[k]
+            if k == "tools":
+                if isinstance(row[k], str) and len(row[k]) > 0:
+                    additional_inputs[k] = json.loads(row[k])
+            else:
+                additional_inputs[k] = row[k]
```
Comment (fabianlim) [inline open_instruct/dataset_transformation.py:1250]:
```suggestion
            row_data = row[k]
            try:
                    row_data = json.loads(row_data)
            except json.decoder.JsonDecodeError:
                     pass
                    
            additional_inputs[k] = row_data
```

File: open_instruct/dataset_transformation.py [hunk 5]
```diff
@@ -1247,6 +1261,8 @@ def masking_strategy_span_search(input_ids: torch.tensor, tokenizer):
         add_generation_prompt=False,
         **additional_inputs,
     )
+    
+    asst_tag = asst_tag + "\n<think>\n" if is_think else asst_tag
```
Comment (fabianlim) [inline open_instruct/dataset_transformation.py:1265]: you can make `<think>` tag passed into this function

Comment (divya-kumari32) [reply open_instruct/dataset_transformation.py:1268]: It wouldn't make a difference I believe and the function's purpose is just to determine think sample so it could be confusing

File: open_instruct/dataset_transformation.py [hunk 6]
```diff
@@ -1248,6 +1264,8 @@ def masking_strategy_span_search(input_ids: torch.tensor, tokenizer):
         **additional_inputs,
     )
 
+    asst_tag = asst_tag + "\n<think>\n" if is_think else asst_tag
```
Comment (Swanand-Kadhe) [inline open_instruct/dataset_transformation.py:1267]: I think we need `is_think(messages)`
- The following test failed for me:
```
ex = {'messages': [
    {"role": "user", "content": "What's the weather like in Bengaluru?"},
    {"role": "assistant", "content": "It's hot"},
]
}
text = tokenizer.apply_chat_template(
    ex['messages'],
    tokenize=False,
    add_generation_prompt=False,
    add_special_tokens=False,
    thinking=False # Switches between thinking and non-thinking modes. Default is True.
)
# print(text)

out = sft_span_seach_mask_out(ex, tokenizer, 4096)

for idx, val in enumerate(out['labels']):
    if val == -100:
        out['labels'][idx] = 81


print(tokenizer.decode(out['input_ids']))
assert text == tokenizer.decode(out['input_ids'])
print('---------------')
print(tokenizer.decode(out['labels']))
```
Output:
```
<|start_of_role|>user<|end_of_role|>What's the weather like in Bengaluru?<|end_of_text|>
<|start_of_role|>assistant<|end_of_role|>It's hot<|end_of_text|>

---------------
<|start_of_role|>user<|end_of_role|>What's the weather like in Bengaluru?<|end_of_text|>
<|start_of_role|>assistant<|end_of_role|>It's hot<|end_of_text|>
```

Comment (divya-kumari32) [reply open_instruct/dataset_transformation.py]: Just corrected it, thanks!

File: open_instruct/dataset_transformation.py (hunk 2 above)
Comment (fabianlim) [reply open_instruct/dataset_transformation.py:1208]: if you want to have this by default, then pls document accurately why it would not affect other tokenizers

File: open_instruct/dataset_transformation.py [hunk 7]
```diff
@@ -1248,6 +1263,10 @@ def masking_strategy_span_search(input_ids: torch.tensor, tokenizer):
         **additional_inputs,
     )
 
+    if check_sample:
```
Comment (Swanand-Kadhe) [inline open_instruct/dataset_transformation.py:1266]: Minor: I wanted to confirm the purpose of `check_sample` flag. Per my understanding, if `check_sample==True` then depending on `is_think`, `<think>` is added in span search. However, the name `check_sample` sounds a bit confusing to me -- sounds like it is related to some debugging. It will be really helpful to add some documentation and/or appropriately change the name. Thanks!

File: open_instruct/dataset_transformation.py (hunk 2 above)
Comment (fabianlim) [reply open_instruct/dataset_transformation.py:1209]: Question, what if the `_asst_tag` or the `_end_tag` is a special token, then what will `tokenizer.encode(end_tag, add_special_tokens=False)` produce?

File: open_instruct/dataset_transformation.py (hunk 7 above)
Comment (fabianlim) [reply open_instruct/dataset_transformation.py]: we keep `mask_think_tag=True` by default because that is the legacy behavior.. am I right @Swanand-Kadhe ?

File: open_instruct/dataset_transformation.py [hunk 8]
```diff
@@ -1180,24 +1180,40 @@ def sft_span_seach_mask_out(
     max_seq_length: int,
     asst_tag: str = "<|start_of_role|>assistant<|end_of_role|>",
     end_tag: str = "<|end_of_text|>",
+    think_tag: str = "\n<think>\n",
+    mask_think_tag: bool = False,
```
Comment (fabianlim) [inline open_instruct/dataset_transformation.py:1184]: this should be `True` by default I believe

**Review by fabianlim — CHANGES_REQUESTED** [review]
Can you accept the two suggestions

File: open_instruct/dataset_transformation.py [hunk 9]
```diff
@@ -1248,6 +1267,12 @@ def masking_strategy_span_search(input_ids: torch.tensor, tokenizer):
         **additional_inputs,
     )
 
+    # if user has specified appendind think tag to base asst tag and the specific row's message column is a thinking sample,
+    # then <think> token is to be included in the asst_tag
```
[the remark this answers is not in the document]
Comment (divya-kumari32) [reply open_instruct/dataset_transformation.py]: Just made the change, thanks

File: open_instruct/dataset_transformation.py [hunk 10]
```diff
@@ -1248,6 +1267,13 @@ def masking_strategy_span_search(input_ids: torch.tensor, tokenizer):
         **additional_inputs,
     )
 
+    # If the user has set `append_think_tag=True` and the current sample is a thinking sample,
```
Comment (fabianlim) [inline open_instruct/dataset_transformation.py:1270]: ok thanks for the comment update.. so can we then change `append_think_tag` to `mask_think_tag_if_present`? Would that be more clear?

File: open_instruct/dataset_transformation.py [hunk 11]
```diff
@@ -1248,6 +1267,13 @@ def masking_strategy_span_search(input_ids: torch.tensor, tokenizer):
...
+    # then the <think> token is appended to the base `asst_tag` used for span matching.
+    # This causes the <think> tag to be masked along with the asst_tag 
+    if append_think_tag:
+        if has_thinking_content(messages):
+            asst_tag += think_tag
+
```
[the remark this answers is not in the document]
Comment (divya-kumari32) [reply open_instruct/dataset_transformation.py:1276]: I didn't put it together because I don't want the code to unnecessarily call the function if the user has explicitly mentioned not to use think tags.

File: open_instruct/dataset_transformation.py (hunk 10 above)
Comment (divya-kumari32) [reply open_instruct/dataset_transformation.py:1270]: I think it is pretty clear with the name, I don't want to unecessarily add too many words as it might get confusing

**Review by fabianlim — APPROVED** [review]
